cdc/owner/async_sink.go

```diff
@@ -46,82 +47,99 @@ type AsyncSink interface {
 	// the DDL event will be sent to another goroutine and execute to downstream
 	// the caller of this function can call again and again until a true returned
 	EmitDDLEvent(ctx cdcContext.Context, ddl *model.DDLEvent) (bool, error)
-	SinkSyncpoint(ctx cdcContext.Context, checkpointTs uint64) error
+	EmitSyncPoint(ctx cdcContext.Context, checkpointTs uint64) error
 	Close(ctx context.Context) error
 }
 
 type asyncSinkImpl struct {
-	sink           sink.Sink
-	syncpointStore sink.SyncpointStore
+	lastSyncPoint  model.Ts
+	syncPointStore sink.SyncpointStore
 
-	checkpointTs model.Ts
-
-	lastSyncPoint model.Ts
-
-	ddlCh         chan *model.DDLEvent
+	checkpointTs  model.Ts
 	ddlFinishedTs model.Ts
 	ddlSentTs     model.Ts
 
+	ddlCh chan *model.DDLEvent
+	errCh chan error
+
+	sink            sink.Sink
+	sinkInitHandler asyncSinkInitHandler
+
 	cancel context.CancelFunc
-	wg     sync.WaitGroup
-	errCh  chan error
 }
 
-func newAsyncSink(ctx cdcContext.Context) (AsyncSink, error) {
-	ctx, cancel := cdcContext.WithCancel(ctx)
-	changefeedID := ctx.ChangefeedVars().ID
-	changefeedInfo := ctx.ChangefeedVars().Info
-	filter, err := filter.NewFilter(changefeedInfo.Config)
+func newAsyncSink() asyncSink {
+	return &asyncSinkImpl{
+		ddlCh:           make(chan *model.DDLEvent, 1),
+		errCh:           make(chan error, defaultErrChSize),
+		sinkInitHandler: asyncSinkInitializer,
+		cancel:          func() {},
+	}
+}
+
+type asyncSinkInitHandler func(ctx cdcContext.Context, a *asyncSinkImpl) error
+
+func asyncSinkInitializer(ctx cdcContext.Context, a *asyncSinkImpl) error {
```

Comment:
May be it is better to take the required variables out of ctx and pass them in through parameters, and make ctx only use for cancel.